openid-4-verifiable-presentations-1_0.md

```diff
@@ -549,11 +555,276 @@ The Wallet then validates the request as specified in OAuth 2.0 [@RFC6749].
+apply to the metadata and validity data of the Credential. The properties of
+this object are defined per Credential Format in (#format_specific_properties). If omitted,
+no specific constraints are placed on the metadata or validity of the requested
+Credential.
+
+`expect_claims`:
+: OPTIONAL. A non-empty array of objects as defined in (#claims_query) that specifies
+claims in the requested Credentials.
+
+`valid_claim_sets`:
+: OPTIONAL. A non-empty array containing arrays of identifiers for elements in
+`expect_claims`. The identifier MAY be postfixed by `?`, indicating that
+delivery of the the respective claim is OPTIONAL.  
+
+## Claims Query {#claims_query}
+
+Each entry in `expect_claims` MUST be an object with the following properties:
+
+`id`:
+: REQUIRED if `valid_claim_sets` is present; OPTIONAL otherwise. A string
+identifying the particular claim. The value MUST be a non-empty string
+consisting of alphanumeric, underscore (`_`) or hyphen (`-`) characters.
+Within the particular `expected_claims` array, the same `id` MUST NOT
+be present more than once.
+
+`path`:
+: REQUIRED if the Credential Format uses a JSON-based claims structure; MUST NOT
+be present otherwise. The value MUST be a claims path pointer that specifies the path to the claim
+within the Verifiable Credential, as defined in (#claims_path_pointer).
+
+`namespace`:
+: REQUIRED if the Credential Format is based on mdoc format defined in ISO 18013-5; MUST NOT be present otherwise. 
+The value MUST be a string that specifies the namespace of the claim
+within the Verifiable Credential, e.g., `org.iso.18013.5.1`.
+
+`claim_name`:
+: REQUIRED if the Credential Format is based on mdoc format defined in ISO 18013-5; MUST NOT be present otherwise. 
+The value MUST be a string that specifies the name of the claim within the provided namespace
+in the Verifiable Credential, e.g., `first_name`.
+
+`value`:
+: OPTIONAL. A string, integer or boolean value that specifies the expected value of the claim. If the
+`value` property is present, the Wallet MUST return the claim only if the type and value
+of the claim match the type and value specified in the query.
+
+`values`:
+: OPTIONAL. An array of strings, integers or boolean values that specifies the expected values of the claim.
+If the `values` property is present, the Wallet MUST return the claim only if the
+type and value of the claim both match for at least one of the elements in the array.
+
+### Selecting Claims and Credentials {#vp_query_lang_processing_rules}
+
+The same basic logic applies for selecting claims and for selecting credentials, as detailed in the following.
+
+#### Selecting Claims
+
+The following rules apply for selecting claims via `expect_claims` and `valid_claim_sets`:
+
+- If `expect_claims` is not provided, the Verifier requests all claims existing
+  in the Credential.
+- If `expect_claims` is provided, but `valid_claim_sets` is not provided,
```

Comment:
Does this mean that each potential combination of claims must be addressed. Does this mean that for each element where there are N alternatives, the amount of option in valdi_claim_sets is multiplied by (N+1)?
e.g. if you request
first_name_latin OR first_name_local
last_name_latin OR last_name_local
age_18 OR age_21 OR age_years OR date_of_birth
street_name OR full_address

You would have 24 sets in valid_claim_sets?

Reply:
That is correct with the current design.

Reply:
The main limitation beyond verbosity I see with this approach is that if an RP wanted to convey say a purpose for requesting a piece of information from a user that maybe sourced from one or multiple possible claims there wouldn't be a way to do that efficiently. For example in the above an RP might want a single purpose string of "Your first name for our customer record" which would be common to both "first_name_latin" and "first_name_local" but I don't see how we could express that efficiently.

Reply:
in query language option 1, there was no need for this kind of duplication, because the nested structure was used and people did not like that. The good news is that all that you mention can be expressed: ["first_name_latin" and "customer_record"] or ["first_name_local" and "customer_record"]. So if the preference is to not use nested structure, we should stick to this syntax, even if it means a lot of enumerations for certain cases. It could probably be optimized, but not sure we have enough reasons to start optimizing now.

Reply:
I'd like to add that you get repetitions for both "clean" forms, [CNF](https://en.wikipedia.org/wiki/Conjunctive_normal_form) and [DNF](https://en.wikipedia.org/wiki/Disjunctive_normal_form). There's no way around that.